package/MDAnalysis/analysis/Aggregation_size.py

```diff
@@ -0,0 +1,370 @@
+                    backend='OpenMP')
+                d[i][ii] = np.amin(dists)
+                ii += 1
+
+    def center_of_mass(self, d):
+        """
+        This function calculates the distance between the COM
+        of two molecules.
+        """
+        self._cut_off = self._cut_off or 8.5
+        cm = {}
+        for x in range((self._molnum)):
+            cm[x+1] = u.select_atoms(f"resid {x+1}").center_of_mass()
+        cm_coordintes = np.array(list(cm.values()))
+        dists = distances.distance_array(
+            cm_coordintes, cm_coordintes, box=u.universe.dimensions)
+        for j in range(self._molnum):
+            ii = 0
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def specific_atoms(self, d):
+        """
+        This function calculates the distance between the selected atom
+        of two molecules.
+        """
+        if (not(atom_1 or atom_2)):
+            raise ValueError("enter valid atom_1 and atom_2")
+        self._cut_off = self._cut_off or 8.5
+        residue1 = u.select_atoms(f"name {atom_1}").positions
+        residue2 = u.select_atoms(f"name {atom_2}").positions
+        dists = distances.distance_array(
+            residue1, residue2, box=u.universe.dimensions)
+        for j in range(self._molnum):
+            ii = 0
+            for i in range(j+1, self._molnum):
+                d[j+1][ii] = dists[i][j]
+                ii += 1
+
+    def _prepare(self):
+        """
+        Create array of zeroes as a placeholder for differnet parameters.
+        """
+        self.results = np.zeros((self.n_frames, 5))
+        self.d = [0] * self._molnum
+        i = 0
+        for it in reversed(range(1, self._molnum+1)):
+            self.d[i] = np.zeros(it * 1).reshape(it, 1)
+            i += 1
+        self.size_num = [0] * self.n_frames
+
+    def aggregation(self, k, tss, counter, m, cut_off, connected):
+        """
+        This is a subfunction that enables the code to find all
+        connected molecules in a cluster.
+        k is a connected molecule to the main molecule in the first level.
+        self._frame_index is a time frame.
+        counter is a step.
+        m provides a suitable off_set to match the number of molecules.
+        Limit is the cut_off provided by the user.
+        """
+        bool_sector = self.d[k] < cut_off
+        Coordinates = self.d[k][bool_sector]
+        # This "if" will check if there is any uninvestigated molecule left
+        if len(Coordinates) != 0:
+            index1 = np.where(bool_sector)
+            listOfCoordinates = list(zip(index1[0], index1[1]))
+            connect = ([j[0] for j in listOfCoordinates]
+                       + connected[m] + 1)
+            Connected_NLevel = np.array(connect)
+        else:
+            Connected_NLevel = []
+        return Connected_NLevel
+
+    def _single_frame(self):
```

Comment:
This function is complicated and long. See if you can break it up into logical units and use methods for different parts to bring out the logic.

I also see lists and various casts with `map` and wonder if you are using the right data structures.